### PR TITLE
Make BeamSearchDecoder compatible with float16

### PR DESCRIPTION
**Pull request by georgesterpu**
Make BeamSearchDecoder compatible with float16 when a non-zero `length_penalty_weight` is used.
Addresses #22172

### CODE REVIEW

File: tensorflow/contrib/seq2seq/python/ops/beam_search_decoder.py [hunk 1]
```diff
@@ -819,6 +819,8 @@ def _get_scores(log_probs, sequence_lengths, length_penalty_weight):
   """
   length_penalty_ = _length_penalty(
       sequence_lengths=sequence_lengths, penalty_factor=length_penalty_weight)
+  if not isinstance(length_penalty_, float):
```
[the remark this answers is not in the document]
Comment (georgesterpu) [reply tensorflow/contrib/seq2seq/python/ops/beam_search_decoder.py]: Hi Eugene.
When `length_penalty_` is a Python float, math_ops.cast will cast it into a Tensor. Would this add one more node to the graph, making the default case slightly slower ?
Would it be better to import and use gen_math_ops.cast instead, plus the type checking ?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yifeif: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yifeif: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

File: tensorflow/contrib/seq2seq/python/ops/beam_search_decoder.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ebrevdo) [reply tensorflow/contrib/seq2seq/python/ops/beam_search_decoder.py]: it gets converted to a float32 tensor anyway somewhere.  use cast, or use ops.convert_to_tensor(length_penalty_, log_probs.dtype) if you want it to error out if someone provides a float32 *tensor* penalty but does float16 calculations.  that's probably better anyway.

[the remark this answers is not in the document]
Comment (georgesterpu) [reply tensorflow/contrib/seq2seq/python/ops/beam_search_decoder.py]: Thanks for the reply, I'll try it.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @yifeif: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Pull request closed without merging**